### PR TITLE
User profile pages

### PR DESCRIPTION
**Pull request by sc-arecrow**
# Possible TODOs:
* [ ] Display photos
* [ ] Settings

### CODE REVIEW

File: ChopChop/Storage/FirebaseDatabase.swift [hunk 1]
```diff
@@ -65,14 +65,26 @@ struct FirebaseDatabase {
             .eraseToAnyPublisher()
     }
 
-    func fetchAllUsers() -> AnyPublisher<[User], Error> {
+    func fetchAllUsers(query: String = "") -> AnyPublisher<[User], Error> {
```
Comment (chrystalquek) [inline ChopChop/Storage/FirebaseDatabase.swift:68]: just a tiny note on an efficiency problem that we cannot solve:
firebase dosent allow us to do search by string on server side so each time we are actually fetching all users and doing front end filtering

File: ChopChop/ViewModels/OnlineRecipes/OnlineRecipeViewModel.swift [hunk 2]
```diff
@@ -43,6 +44,8 @@ class OnlineRecipeViewModel: ObservableObject {
             .sink { [weak self] recipe in
                 self?.recipe = recipe
 
+                self?.recipeServingText = "\(recipe.servings.removeZerosFromEnd()) \(recipe.servings == 1 ? "person" : "people")"
```
Comment (chrystalquek) [inline ChopChop/ViewModels/OnlineRecipes/OnlineRecipeViewModel.swift:47]: actually this particular format is used in all recipes, maybe we can generalise it as view component somewhere

File: ChopChop/ViewModels/Users/ProfileViewModel.swift [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+    }
+
+    private var userPublisher: AnyPublisher<User, Never> {
+        storageManager.userByIdPublisher(userId: userId)
+            .assertNoFailure()
+            .eraseToAnyPublisher()
+    }
+
+    private static func getRecipesPublisher(userId: String) -> AnyPublisher<[OnlineRecipe], Error> {
+        StorageManager().allRecipesByUsersPublisher(userIds: [userId])
```
Comment (chrystalquek) [inline ChopChop/ViewModels/Users/ProfileViewModel.swift:82]: maybe can set StorageManager as a private variable

File: ChopChop/Views/Users/AddFolloweeView.swift [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+
+struct AddFolloweeView: View {
+    @ObservedObject var viewModel: NonFolloweeCollectionViewModel
+
+    var body: some View {
+        VStack {
+            SearchBar(text: $viewModel.query, placeholder: "Search...")
+
+            if viewModel.nonFollowees.isEmpty {
+                NotFoundView(entityName: "Non Followees")
```
Comment (chrystalquek) [inline ChopChop/Views/Users/AddFolloweeView.swift:11]: maybe "users"? otherwise it will show "No Non Followees Found"

**Review by chrystalquek — APPROVED** [review]

File: ChopChop/Storage/FirebaseDatabase.swift (hunk 1 above)
Comment (sc-arecrow) [reply ChopChop/Storage/FirebaseDatabase.swift:68]: Yes I think this is the best we can achieve with Firebase restrictions

File: ChopChop/ViewModels/Users/ProfileViewModel.swift (hunk 3 above)
Comment (sc-arecrow) [reply ChopChop/ViewModels/Users/ProfileViewModel.swift:82]: Issue with that is that this publisher is required before every property in the view model is initialised, so using `self.storageManager` is not allowed

**Pull request merged**